pkg/actuators/awsplacementgroup/controller.go

```diff
@@ -0,0 +1,103 @@
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	"k8s.io/klog/v2"
+)
+
+// Reconciler reconciles machineSets.
+type Reconciler struct {
+	Client client.Client
+	Log    logr.Logger
+
+	recorder record.EventRecorder
+	scheme   *runtime.Scheme
+}
+
+// SetupWithManager creates a new controller for a manager.
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+	_, err := ctrl.NewControllerManagedBy(mgr).
+		For(&machinev1.AWSPlacementGroup{}).
+		WithOptions(options).
+		Build(r)
+
+	if err != nil {
+		return fmt.Errorf("failed setting up with a controller manager: %w", err)
+	}
+
+	r.recorder = mgr.GetEventRecorderFor("awsplacementgroup-controller")
+	r.scheme = mgr.GetScheme()
+	return nil
+}
+
+// Reconcile implements controller runtime Reconciler interface.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := r.Log.WithValues("awsplacementgroup", req.Name, "namespace", req.Namespace)
+	logger.V(3).Info("Reconciling")
+
+	awsPlacementGroup := &machinev1.AWSPlacementGroup{}
+	if err := r.Client.Get(ctx, req.NamespacedName, awsPlacementGroup); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return. Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// Ignore deleted AWSPlacementGroups, this can happen when foregroundDeletion
+	// is enabled
+	if !awsPlacementGroup.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+	originalAWSPlacementGroupToPatch := client.MergeFrom(awsPlacementGroup.DeepCopy())
+
+	result, err := r.reconcile(awsPlacementGroup)
+	if err != nil {
+		logger.Error(err, "Failed to reconcile AWSPlacementGroup")
+		r.recorder.Eventf(awsPlacementGroup, corev1.EventTypeWarning, "ReconcileError", "%v", err)
+		// we don't return here so we want to attempt to patch the AWSPlacementGroup regardless of an error.
+	}
+
+	if err := r.Client.Patch(ctx, awsPlacementGroup, originalAWSPlacementGroupToPatch); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to patch AWSPlacementGroup: %v", err)
+	}
+
+	if isInvalidConfigurationError(err) {
+		// For situations where requeuing won't help we don't return error.
+		// https://github.com/kubernetes-sigs/controller-runtime/issues/617
+		return result, nil
+	}
+	return result, err
+}
+
+func isInvalidConfigurationError(err error) bool {
+	switch t := err.(type) {
+	case *mapierrors.MachineError:
+		if t.Reason == machinev1beta1.InvalidConfigurationMachineError {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
You could use the `errors.Is` helper that was introduced to the go errors package in 1.16

Reply:
Temporarily removed.